added import/export feature - depended on migration PR

MykhailoDmytriakha opened the pull request:
<img width="297" alt="image" src="https://github.com/user-attachments/assets/e1c31f86-2e2f-4bce-a299-4fa42a79ec89">

<img width="293" alt="image" src="https://github.com/user-attachments/assets/682afd65-8e01-4706-b8bb-aa36872105af">

---

Экспорт/импорт пользовательских данных в формате JSON:

## Описание: Функциональность импорта и экспорта базы данных

Этот pull request добавляет возможность импорта и экспорта пользовательских данных в приложении, позволяя пользователям создавать резервные копии своих данных или переносить их между устройствами.

**Функциональность:**

*   **Экспорт:**
    *   Пользователь может инициировать операцию экспорта через меню настроек приложения.
    *   Приложение генерирует JSON-файл, содержащий данные пользователя, включая:
        *   Настройки приложения (например, размер текста песни, развернут ли текст песни, тема приложения).
        *   Настройки книг (предпочитаемые пользователем книги и связанные с ними значения предпочтений).
        *   Список избранного (сохраненные песни).
        *   Отредактированные песни (песни, измененные пользователем).
        *   Пользовательские категории (созданные пользователем категории, включая их названия, цвета и связанные с ними песни).
    *   Пользователю предлагается выбрать место для сохранения экспортированного JSON-файла с помощью системного диалога выбора файлов.
    *   Временный файл используется для сборки данных экспорта перед записью в конечный пункт назначения, что повышает эффективность обработки данных.
*   **Импорт:**
    *   Пользователь может инициировать операцию импорта через меню настроек приложения.
    *   Приложение предлагает пользователю выбрать ранее экспортированный JSON-файл, используя системный диалог выбора файлов.
    *   Приложение анализирует JSON-файл и импортирует данные, перезаписывая существующие данные, где это применимо:
        *   Обновляются настройки приложения.
        *   Настройки книг обновляются или создаются, если они не существуют.
        *   Добавляются избранные.
        *   Отредактированные песни обновляются или добавляются.
        *   Пользовательские категории обновляются или создаются вместе со связанными с ними песнями.
    *   Приложение предоставляет пользователю обратную связь об успехе или неудаче операции импорта/экспорта.

**Детали реализации:**

*   **`ImportExportPreferenceFragment`:** Новый подкласс `PreferenceFragmentCompat`, отвечающий за обработку элементов пользовательского интерфейса и взаимодействие с пользователем, связанное с операциями импорта и экспорта. Он использует `ActivityResultContracts` для взаимодействия с системным диалогом выбора файлов для выбора и сохранения файлов.
*   **`ImportExportViewModel`:** Подкласс `ViewModel`, который обрабатывает логику экспорта и импорта данных. Он взаимодействует с базой данных через DAO и использует сопрограммы для фоновых операций.
*   **Сериализация данных:** `jacksonObjectMapper()` из модуля Jackson Kotlin используется для сериализации и десериализации пользовательских данных в формат JSON и обратно.
*   **Модели данных:**
    *   `ExportData`: Представляет общую структуру экспортируемых данных.
    *   `Preferences`: Представляет настройки приложения и книг пользователя.
    *   `BookPreference`: Представляет предпочтения пользователя для определенной книги.
    *   `CategoryDetails`: Представляет созданную пользователем категорию и связанные с ней песни.
*   **Обработка ошибок:** Процессы импорта и экспорта обернуты в блоки `try-catch` для обработки потенциальных исключений и предоставления соответствующей обратной связи пользователю.
*   **Логирование:** Timber используется для регистрации важных событий и ошибок во время процесса импорта/экспорта.

**Преимущества:**

*   **Резервное копирование данных:** Пользователи могут создавать резервные копии своих данных, чтобы предотвратить потерю данных.
*   **Передача данных:** Пользователи могут легко переносить свои данные на новое устройство.
*   **Переносимость данных:** Использование JSON в качестве формата экспорта гарантирует, что данные могут быть легко переданы и потенциально использованы с другими приложениями.

**Примечание:**

*   Эта функция предназначена для обработки текущей схемы базы данных. Однако ее можно адаптировать для поддержки будущих изменений схемы, включая введение новых системных сборников песен и пользовательских сборников песен.
*   **Обработка пользовательских сборников песен и обновлений схемы:** Функциональность импорта/экспорта может способствовать бесшовному обновлению даже при наличии пользовательских сборников песен. Например, если в будущем обновлении будет добавлен новый системный сборник песен, можно использовать следующий процесс:
    1. **Экспорт пользовательских данных:** Текущие данные пользователя, включая пользовательские сборники песен и любые изменения существующих песен, экспортируются в JSON-файл.
    2. **Применение миграций базы данных:** База данных приложения обновляется, чтобы включить новый системный сборник песен и любые другие изменения схемы. Это гарантирует, что структура базы данных готова к новому контенту.
    3. **Импорт пользовательских данных:** Ранее экспортированный JSON-файл импортируется. Во время импорта:
        *   **Приоритет системных сборников песен:** Системные сборники песен рассматриваются как источник истины.
        *   **Объединение пользовательских данных:** Пользовательские сборники песен, правки песен и другие настройки тщательно объединяются с обновленной базой данных:
            *   Добавляются пользовательские сборники песен.
            *   Пользовательские правки системных песен, которые также присутствуют в новом системном сборнике песен, будут обрабатываться на основе предопределенной стратегии разрешения конфликтов (например, приоритет пользовательских правок, более новых правок или предоставление пользователю возможности выбора).
            *   Другие пользовательские данные (избранное, настройки) импортируются как обычно.
*   Использование временных файлов во время экспорта обеспечивает эффективную обработку данных, особенно для больших наборов данных.
*   Использование системного средства выбора файлов обеспечивает согласованный и удобный способ управления файлами.

Этот pull request расширяет функциональность приложения, предоставляя пользователям больший контроль над своими данными и повышая общее удобство использования.
---

export/import user data to JSON feature:
## Feature: Database Import/Export Functionality

This pull request introduces the ability to import and export user data within the application, allowing users to back up their data or transfer it between devices.

**Functionality:**

*   **Export:**
    *   The user can initiate an export operation through the app's settings menu.
    *   The app generates a JSON file containing the user's data, including:
        *   App preferences (e.g., song text size, song text expanded, app theme).
        *   Book preferences (user's preferred books and their associated preference values).
        *   Favorites list (saved songs).
        *   Edited songs (songs modified by the user).
        *   Custom categories (user-created categories, including their names, colors, and associated songs).
    *   The user is prompted to choose a location to save the exported JSON file using the system's file picker.
    *   A temporary file is used to build the export data before writing it to the final destination, improving data handling efficiency.
*   **Import:**
    *   The user can initiate an import operation through the app's settings menu.
    *   The app prompts the user to select a previously exported JSON file using the system's file picker.
    *   The app parses the JSON file and imports the data, overwriting existing data where applicable:
        *   App preferences are updated.
        *   Book preferences are updated or created if they don't exist.
        *   Favorites are added.
        *   Edited songs are updated or added.
        *   Custom categories are updated or created, along with their associated songs.
    *   The app provides feedback to the user about the success or failure of the import/export operation.

**Implementation Details:**

*   **`ImportExportPreferenceFragment`:** A new `PreferenceFragmentCompat` subclass responsible for handling the UI elements and user interactions related to import and export operations. It uses `ActivityResultContracts` to interact with the system's file picker for selecting and saving files.
*   **`ImportExportViewModel`:** A `ViewModel` subclass that handles the logic for exporting and importing data. It interacts with the database through DAOs and utilizes coroutines for background operations.
*   **Data Serialization:** The `jacksonObjectMapper()` from the Jackson Kotlin module is used to serialize and deserialize the user data to and from JSON format.
*   **Data Models:**
    *   `ExportData`: Represents the overall structure of the exported data.
    *   `Preferences`: Represents the user's app and book preferences.
    *   `BookPreference`: Represents the user's preference for a specific book.
    *   `CategoryDetails`: Represents a user-created category and its associated songs.
*   **Error Handling:** The import and export processes are wrapped in `try-catch` blocks to handle potential exceptions and provide appropriate feedback to the user.
*   **Logging:** Timber is used to log important events and errors during the import/export process.

**Benefits:**

*   **Data Backup:** Users can create backups of their data to prevent data loss.
*   **Data Transfer:** Users can easily transfer their data to a new device.
*   **Data Portability:** The use of JSON as the export format ensures that the data can be easily shared and potentially used with other applications.

**Note:**

*   This feature is designed to handle the current database schema. However, it can be adapted to support future schema changes, including the introduction of new system songbooks and custom user songbooks.
*   **Handling Custom Songbooks and Schema Updates:** The import/export functionality can facilitate seamless updates even when custom songbooks are present. For instance, if a new system songbook is added in a future update, the following process can be employed:
    1. **Export User Data:** The user's current data, including custom songbooks and any edits to existing songs, is exported to a JSON file.
    2. **Apply Database Migrations:** The app's database is updated to include the new system songbook and any other schema changes. This will ensure that the database structure is ready for the new content.
    3. **Import User Data:** The previously exported JSON file is imported. During the import:
        *   **Prioritize System Songbooks:** The system songbooks are treated as the source of truth.
        *   **Merge User Data:** The user's custom songbooks, song edits, and other preferences are carefully merged with the updated database:
            *   Custom songbooks are added.
            *   User edits to system songs that are also present in the new system songbook will be handled based on a predefined conflict resolution strategy (e.g., prioritize user edits, newer edits, or provide an option to the user to choose).
            *   Other user data (favorites, preferences) are imported as usual.
*   The use of temporary files during export ensures efficient data handling, especially for large datasets.
*   Using the system's file picker provides a consistent and user-friendly way to manage files.

This pull request enhances the app's functionality by providing users with greater control over their data and improving its overall usability.

Commit:
added import/export feature

alelk commented:
У нас появляется внешний контракт, было бы хорошо покрыть тестами (Kotest) результирующий json. Если позднее модель будет меняться, нужно будет быть уверенным, что сохранилась возможность распарсить предыдущие версии модели. Так же желательно добавить поле "version" в экспортируемый json (Например, "version": 1).

Commit:
test: add comprehensive tests for ExportData serialization

- Add tests to verify version field inclusion in exported JSON
- Ensure backward compatibility with JSON files without version field
- Validate correct format and presence of all required fields
- Test proper serialization/deserialization of complex types
- Add real-world JSON example test case

This ensures stable JSON format and maintains backward compatibility
for the import/export functionality.

MykhailoDmytriakha commented:
Here's what was added:

1. Added a version field to the `ExportData` class with a default value of 1
2. Created comprehensive tests that verify:
   - The version field is included in the exported JSON
   - Backward compatibility with JSON files that don't have a version field
   - The correct format of the exported JSON, including all required fields
   - Proper serialization and deserialization of all data types, including colors
   - Proper handling of null values

The tests ensure that:
1. The JSON format is stable and well-defined
2. Future changes to the model will maintain backward compatibility
3. All fields are properly serialized and deserialized
4. The version field is always present in new exports but optional when parsing

Commit:
ci: enhance test coverage and reporting

- Add separate step for running unit tests for both ru/uk flavors
- Add test results upload to artifacts for better debugging
- Keep test reports for 14 days in GitHub Actions
- Add stacktrace flag for better error reporting

The original build process remains unchanged while adding more comprehensive
test coverage and visibility into test results.

MykhailoDmytriakha commented:
updated workflow:

- Add separate step for running unit tests for both ru/uk flavors
- Add test results upload to artifacts for better debugging
- Keep test reports for 14 days in GitHub Actions
- Add stacktrace flag for better error reporting

The original build process remains unchanged while adding more comprehensive
test coverage and visibility into test results.

MykhailoDmytriakha commented:
PR is updated with required changes.
https://github.com/alelk/pws-android/pull/40


On Wed, Dec 18, 2024 at 1:08 AM Alex Elkin ***@***.***> wrote:

> У нас появляется внешний контракт, было бы хорошо покрыть тестами (Kotest)
> результирующий json. Если позднее модель будет меняться, нужно будет быть
> уверенным, что сохранилась возможность распарсить предыдущие версии модели.
> Так же желательно добавить поле "version" в экспортируемый json (Например,
> "version": 1).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/alelk/pws-android/pull/40#issuecomment-2550760706>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AN4RYZNITARAJX7FOX3FQTD2GE3R3AVCNFSM6AAAAABTJLMZMWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKNJQG43DANZQGY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
